i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,245 @@
+from __future__ import annotations
+
+__all__ = ["ConformerMHSARelPosV1", "ConformerMHSARelPosV1Config"]
+
+from dataclasses import dataclass
+import math
+from typing import Optional, Literal
+
+import torch
+from torch import nn
+import torch.nn.functional as F
+
+from i6_models.config import ModelConfiguration
+from i6_models.util import compat
+from i6_models.parts.dropout import BroadcastDropout
+
+
+@dataclass
+class ConformerMHSARelPosV1Config(ModelConfiguration):
+    """
+    Attributes:
+        input_dim: input dim and total dimension for query/key and value projections, should be divisible by `num_att_heads`
+        num_att_heads: number of attention heads
+        with_bias: whether to add bias to qkv and output linear projections
+        att_weights_dropout: attention weights dropout
+        learnable_pos_emb: whether to use learnable relative positional embeddings instead of fixed sinusoidal ones
+        rel_pos_clip: maximal relative postion for embedding
+        with_linear_pos: whether to linearly transform the positional embeddings
+        separate_pos_emb_per_head: whether to create head-dependent positional embeddings
+        with_pos_bias: whether to add additional position bias terms to the attention scores
+        pos_emb_dropout: dropout for the positional embeddings
+        dropout: multi-headed self attention output dropout
+        dropout_broadcast_axes: string of axes to which dropout is broadcast, e.g. "T" for broadcasting to the time axis
+                                setting to None to disable broadcasting
+    """
+
+    input_dim: int
+    num_att_heads: int
+    with_bias: bool
+    att_weights_dropout: float
+    learnable_pos_emb: bool
+    rel_pos_clip: Optional[int]
+    with_linear_pos: bool
+    with_pos_bias: bool
+    separate_pos_emb_per_head: bool
+    pos_emb_dropout: float
+    dropout: float
+    dropout_broadcast_axes: Optional[Literal["B", "T", "BT"]] = None
```

Comment:
I wonder a bit about the choice where we allow defaults and where not. So you have a default for `dropout_broadcast_axes`, but not for others. Why?

Reply:
I don't exactly understand.

> because for `ConformerPositionwiseFeedForwardV1Config` there is already a default value set for `activation`

Why is that relevant here?

> I think dataclass attributes without default can't follow a attribute with a default.

Yes, so just reorder them?

> But I can make the config which doesn't inherit from the V1 config if you prefer no default.

You don't inherit the config here?

I'm not saying you should not have any defaults. Actually the opposite. I like to have reasonable defaults. But as I understood, we don't want to have any defaults in i6_models? Or did we already weaken this rule? So what is really the rule now? I just don't understand that you have a default for `dropout_broadcast_axes` but not for the others.

Reply:
I don't know how the rules are like.

Reply:
from https://github.com/rwth-i6/i6_models/blob/main/CONTRIBUTING.md

```
default values for Part/Assembly config:
* defaults for common hyper-parameters (e.g. number of layers): no
* defaults for rarely/never tuned hyper-parameters (e.g. optimizer constants): maybe
* defaults for structural elements (parts / assemblies / PyTorch modules/functions): yes
* if a PR contains a default (not covered by the third point) there should be a reason given for it in the PR / code
```

Reply:
So, what about the `dropout_broadcast_axes` then? We don't know really yet what variant is the best. Or do we? So no default?

None of the settings here are really structural. (Or maybe the learned pos enc is?)

We also don't really know what's best. Probably people will rarely tune it and just use the common hyper-parameters, so it's a no/maybe.